[16.0][IMP]helpdesk_mgmt: simplify email templates

PicchiSeba opened the pull request:
Forward-ports https://github.com/OCA/helpdesk/pull/467 which itself partially forward-ported 98e5ef7f4a2e074283da3674654f662a28f02eed and dfbc3698e76b3a1330f339ca78b49557b78c05ed from 12.0 to 14.0

eLBati committed:
[IMP] helpdesk_mgmt: Use email layout to simplify all templates
- Avoid email_cc (error "message without recipients")
- Use coherent partner/user reference
- Access button from e-mail notifications

Semi-forward port of 98e5ef7f4a2e074283da3674654f662a28f02eed

Please note that the commit a5041c25a79cd9620d30cc09dd9741f46aef7e76 (An e-mail is sent to the ticket after each create/edit) has not been forward ported because the assignment notify is done internally by odoo in `_message_auto_subscribe_followers` from `mail.thread`

eLBati committed:
[IMP] helpdesk_mgmt: Test access added

Forward port of dfbc3698e76b3a1330f339ca78b49557b78c05ed

PicchiSeba commented:
I just rebased because of a conflict arising from the merge of #643

pedrobaeza commented on helpdesk_mgmt/controllers/main.py:
```diff
@@ -82,6 +82,7 @@ def _prepare_submit_ticket_vals(self, **kw):
             "channel_id": request.env.ref(
                 "helpdesk_mgmt.helpdesk_ticket_channel_web", False
             ).id,
+            "stage_id": request.env.ref("helpdesk_mgmt.helpdesk_ticket_stage_new").id,
```
We shouldn't force the initial stage.

pedrobaeza commented on helpdesk_mgmt/data/helpdesk_data.xml:
```diff
@@ -32,105 +32,53 @@
-                    </tbody>
-                </table>
+                <div>
+                    <p>Hello ${object.partner_id.name or object.partner_name or ''},</p>
```
This is not 16.0 syntax

pedrobaeza commented on helpdesk_mgmt/data/helpdesk_data.xml:
```diff
@@ -32,105 +32,53 @@
+                </div>
+            </field>
+        </record>
+        <record id="created_ticket_template" model="mail.template">
```
This is a new template.

PicchiSeba replied:
Would you prefer its own commit/PR?

pedrobaeza replied:
Or its own PR explaining the reasons, so that not everything must be approved together. Anyway, this is not working, as the syntax is not v16. Did you check it?

PicchiSeba replied:
yes, I'm working on it.

pedrobaeza reviewed (changes requested):
There's an unrelated commit, and the fw-port seems done without checking.

PicchiSeba commented:
I removed the unrelated commit, the new email template (will do another PR for that) and the `stage_id` forced assignment in the controller. I also converted the email templates to the `16.0` format.

Do you see any other critical issue?

HekkiMelody commented on helpdesk_mgmt/tests/test_helpdesk_ticket.py:
```diff
@@ -181,3 +183,101 @@ def test_ticket_without_team(self):
+
+    def test_partner_can_access(self):
+        Ticket = self.env["helpdesk.ticket"]
+        self.assertFalse(Ticket.partner_can_access())
```
**comment (non-blocking)**: This assertion is a bit redundant IMO

HekkiMelody commented on helpdesk_mgmt/tests/test_helpdesk_ticket.py:
```diff
@@ -181,3 +183,101 @@ def test_ticket_without_team(self):
         self.assertEqual(self.new_stage, new_ticket.stage_id)
+
+    def test_helpdesk_ticket_multiple(self):
+        Ticket = self.env["helpdesk.ticket"]
+        Partner = self.env["res.partner"]
```
**suggestion (non-blocking)**: All these could be put in the setUp once, instead of being repeated in every test.

HekkiMelody commented on helpdesk_mgmt/models/helpdesk_ticket.py:
```diff
@@ -176,7 +207,13 @@ def create(self, vals_list):
                 if channel_email_id:
                     vals["channel_id"] = channel_email_id.id
-        return super().create(vals_list)
+        res = super().create(vals_list)
+        created_ticket_template_name = self._get_created_ticket_template_name()
+        if created_ticket_template_name:
+            for rec in res:
+                if rec.partner_id or rec.partner_email:
+                    rec._send_mail_to_partner(created_ticket_template_name)
+        return res
```
**question**: if you have an email template set on the stage, and this global email template on ticket creation, do both emails get sent at the same time?

HekkiMelody commented on helpdesk_mgmt/models/helpdesk_ticket.py:
```diff
@@ -132,6 +132,28 @@ def _read_group_stage_ids(self, stages, domain, order):
+    def _send_mail_to_partner(self, template_name):
+        """Send a feedback message to the contact linked to this newly created ticket
+        (whatever it is a real partner or just its e-mail).
+        Note: The purpose of this function is to replace the stage's `mail_template_id`
+        role (that just doesn't work when the ticket is created from the portal or when
+        `partner_id` is set)"""
```
**issue**: as far as I can tell, so far this doesn't replace but rather adds a new email, no? (see also the other comment below)

HekkiMelody commented:
I see 3 different purposes to this PR
- Refactoring email templates (LGTM on that front)
- Improving how access to tickets is handled for portal users (LGTM on that front too, and it's closely related)
- Hardcoding sending an email on ticket creation no matter what (I have some question, and perhaps this could be split in a different PR?)

PicchiSeba committed:
[IMP]helpdesk_mgmt: simplify tests structure
